src/sagemaker/predictor.py

```diff
@@ -114,18 +114,35 @@ def _delete_endpoint_config(self):
         self.sagemaker_session.delete_endpoint_config(endpoint_config_name)
 
     def delete_endpoint(self, delete_endpoint_config=True):
-        """Delete the Amazon SageMaker endpoint and endpoint configuration backing this predictor.
+        """Delete the Amazon SageMaker endpoint backing this predictor. Also delete the endpoint configuration attached
+           to it if delete_endpoint_config is True.
 
         Args:
-            delete_endpoint_config (bool): Flag to indicate whether to delete the corresponding SageMaker endpoint
-                configuration tied to the endpoint. If False, only the endpoint will be deleted. (default: True)
+            delete_endpoint_config (bool): Flag to indicate whether to delete endpoint configuration together with
+                endpoint. If False, only endpoint will be deleted. Default: True.
 
         """
         if delete_endpoint_config:
             self._delete_endpoint_config()
 
         self.sagemaker_session.delete_endpoint(self.endpoint)
 
+    def delete_model(self):
+        """Deletes the Amazon SageMaker models backing this predictor.
+
+        """
+        model_names = self._get_model_names()
+        for model_name in model_names:
+            self.sagemaker_session.delete_model(model_name)
+
+    def _get_model_names(self):
+        endpoint_desc = self.sagemaker_session.sagemaker_client.describe_endpoint(EndpointName=self.endpoint)
+        endpoint_config_name = endpoint_desc['EndpointConfigName']
+        endpoint_config = self.sagemaker_session.sagemaker_client.describe_endpoint_config(
+            EndpointConfigName=endpoint_config_name)
+        production_variants = endpoint_config['ProductionVariants']
+        return map(lambda d: d['ModelName'], production_variants)
```

Comment:
It looks like it is possible for multiple models to exist behind an endpoint configuration and endpoint.

Let's make sure this class also has the delete_model functionality as well.
https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/pipeline.py

Here is an example on how it is done:
https://github.com/awslabs/amazon-sagemaker-examples/blob/master/sagemaker-python-sdk/scikit_learn_inference_pipeline/Inference%20Pipeline%20with%20Scikit-learn%20and%20Linear%20Learner.ipynb

We should make sure the deletion of multiple models in the same endpoint configuration is tested in an integ test.

Reply:
Something to consider is that this solution assumes the endpoint configuration exists. It is possible for a user to call delete_endpoint() and that might error out when they call delete_model().

Technically, the model object does still exist within SageMaker. It arguably makes sense to actually delete the endpoint first, then endpoint configuration and then model since they all have dependency on each other. I believe SageMaker allows you to delete them regardless of order, since they snapshot everything, however this problem can still persist if the user deletes the endpoint first.

Reply:
Just to confirm, for `Predictor`, `delete_model()` should delete all the SageMaker models from the list[sagemaker.Model] and delete the model with `name` (if `name` is not None)?

And for the order of deletions, if endpoint/endpoint config is deleted before attempting to delete model, should we say something like "model cannot be deleted because the attached endpoint/endpoint config is already gone"?

Reply:
Let's initialize the model name in the constructor per our in person discussion.

Reply:
We're going to call `describe_endpoint()` and `describe_endpoint_config()` once the `Predictor` gets `endpoint_name` in the constructor, so that we don't restrict the order of deletions to model -> endpoint/endpoint config.

Reply:
Change the existing delete_endpoint() functionality to depend on the endpoint_configuration_name instead of describing it.

https://github.com/aws/sagemaker-python-sdk/blob/d7a2bf6be25e8da5ef0ec0e8273c3f9a8d9053d5/src/sagemaker/predictor.py#L112